# Phylo-Spec

## Content

- [Introduction](##Introduction)
- [Package requirement](##Packagerequirement)
- [Installation environment](##Installationenvironment)
- [Build a phylogenetic tree](###Buildaphylogenetictree)
  - [a. 16S Amplicon](####a. 16S Amplicon)
  - [b. WGS Metagenome](####b. WGS Metagenome)
- [Model training and prediction](###Step 2: Model training and prediction)
  - [a. Normal status](####a. Normal status)
  - [b. Unclassified status](####b. Unclassified status)
  - [c. Five cross validation](####c. Five cross validation)
- [Feature Importance Ranking](###Step 3: Feature Importance Ranking)
- [Run the example with one click](##Run the example with one click)
- [Supplementary](##Supplementary)

## Introduction

Phylo-Spec, a phylogeny-driven deep learning algorithm, integrates microbial richness within a phylogenetic hierarchy and assigns unclassified species to virtual nodes, reducing data misalignment and uncertainty.  Phylo-Spec outperforms existing methods in synthetic and real-world datasets, demonstrating superior accuracy in classifying diseases like IBD, CRC, and T2D, establishing it as a robust tool for microbiome-based health prediction.

## Package requirement

```
torch >= 2.3.1
pandas >= 2.2.2
numpy >= 1.26.4
scikit-learn >= 1.4.2
imbalanced-learn >= 0.12.3
ete3 >= 3.1.3
matplotlib >= 3.7.2
biopython >= 1.83
```

## Installation environment

```
sh init.sh
```

```python
cd Phylo-Spec
```

Then all tools are located at ‘src’ folder:

```
PhyloSpec_train_test.py // For model training and testing
PhyloSpec_cv.py // For model five cross validation
PhyloSpec_Imp.py // For model output feature importance
16S_Phlyogeny.py // For model obtaining 16S rRNA amplicon phylogenetic tree
WGS_Phlyogeny.py // For model obtaining metagenomic phylogenetic trees
```

### Build a phylogenetic tree

This step is optional, the phylogeny tree can either be constructed from representative sequences (e.g. marker gene or amplicon), or provided by users from a NEWICK format file (e.g. for shotgun).

#### a. 16S Amplicon

In Windows, you can directly execute the command. The relevant software and database have been integrated into the software package, including Mafft multiple sequence alignment, FastTree phylogenetic tree construction, and Greengenes (v13) database.The Greengenes database we provide is under "./database/16S".You can specify the output path '-o', the default output path is “./output”

| SampleID | 4371463 | 2250985 | ...  | Group   |
| ------- | ------- | ------- | :--: | ------- |
| sample1 | 0.001   | 0.002   |      | Healthy |
| sample2 | 0       | 0.003   |      | Healthy |
| sample3 | 0       | 0       |      | Disease |

```
python ./src/Phylogeny/16S/16S_Phlyogeny.py -c ./example/16S/example_train.csv -f ./database/16S/gg_13database.fa -o output
```

#### b. WGS Metagenome

We provide phylogenetic trees of metaphlan4 and metaphlan3 databases. You can get the phylogenetic tree Newick file of the relevant species by directly executing the command.The database is in "./database/WGS".

| SampleID | s__Clostridia_bacterium | s__Rothia_mucilaginosa | ...  | Group   |
| ------- | ----------------------- | ---------------------- | :--: | ------- |
| sample1 | 0.001                   | 0.002                  |      | Healthy |
| sample2 | 0                       | 0.003                  |      | Healthy |
| sample3 | 0                       | 0                      |      | Disease |

```
python ./src/Phylogeny/WGS/WGS_Phlyogeny.py -c ./example/WGS/example_train.csv -t ./database/WGS/wgs_mpa4_phylogeny.nwk -o output
```

### Model training and prediction

All files generated by the model are in the “./output” folder.

You can assign microbial signature tables, phylogenetic tree.nwK files, and taxonomic tables by '-c', '-t', and '-taxo'(e.g. unclassified species) respectively. In addition, you can specify the output path of the model '-o'. In the meantime, train, test, and cross-validate the model using --PhyloSpec train, --PhyloSpec test, and --PhyloSpec cv

#### a. Normal status

The normal state is the normal input abundance table(e.g. OTU or Species).

Training process:

```
python ./src/model/PhyloSpec_train_test.py -t ./example/16S/phylogeny.nwk -c ./example/16S/example_train.csv --PhyloSpec train
```

Testing Process:

````
python ./src/model/main_train_test.py -t ./example/16S/phylogeny.nwk -c ./example/16S/example_test.csv --PhyloSpec test
````

#### b. Unclassified status

The species contains unclassified cases (e.g. "xxx_Unclassified_xxx" or "xxx_unclassified_xxx").And you need to enter the "-taxo" taxonomy table additionally.

Saved taxonomy.csv and phylogeny.nwk in "./database"(e.g. 16S and WGS)

| example | s__Clostridia_bacterium | s__unclassified | ...  | Group   |
| ------- | ----------------------- | --------------- | :--: | ------- |
| sample1 | 0.001                   | 0.002           |      | Healthy |
| sample2 | 0                       | 0.003           |      | Healthy |
| sample3 | 0                       | 0               |      | Disease |

The input taxonomy format is

```
Kingdom Phylum  Class   Order   Family  Genus   Species
k__Archaea p__Euryarchaeota c__Methanopyri o__Methanopyrales f__Methanopyraceae g__Methanopyrus s__Methanopyrus_kandleri
k__Archaea p__Euryarchaeota c__Methanobacteria o__Methanobacteriales f__Methanothermaceae g__Methanothermus  s__Methanothermus_fervidus
k__Archaea p__Euryarchaeota c__Methanobacteria o__Methanobacteriales f__Methanobacteriaceae g__Methanothermobacter s__Methanothermobacter_marburgensis
```

Training process:

```
python ./src/model/PhyloSpec_train_test.py -t ./example/Unclassified/phylogeny.nwk -c ./example/Unclassified/example_train.csv -taxo ./example/Unclassified/example_taxonomy.csv --PhyloSpec train
```

Testing Process:

```
python ./src/model/PhyloSpec_train_test.py -t ./example/Unclassified/phylogeny.nwk -c ./example/Unclassified/example_test.csv -taxo ./example/Unclassified/example_taxonomy.csv --PhyloSpec test
```

#### c. Five cross validation

We also provide a one-click run of the five cross validation

```
python ./src/model/PhyloSpec_cv.py -t ./example/CV/phylogeny.nwk -c ./example/CV/example_cv.csv -taxo ./example/CV/example_taxonomy.csv --PhyloSpec cv
```

### Phylo-Spec feature importance

All files generated by the model are in the “./output” folder.

The output file is "Feature_Importance_Score.xlsx", which has 5 sub-tables: leaf node (species) importance, internal node importance, all node importance, node relationship table and node relationship "nwk" format

```
python ./src/model/PhyloSpec_Imp.py
```

## Run the example with one click

For convenience, you can run the processes above by running the example.sh in folder './example'.

```
cd ./example
chmod a+x example.sh
./example.sh
```

## Supplementary

**Synthetic Dataset 1**contains 148 artificial microbiomes(common species)

**Synthetic Dataset 2** contains 148 artificial microbiomes(unclassified species)

**Real Dateset 1** contains 2010 IBD 16S amplicon samples processed by Parallel-Meta Suite.

**Real Dateset 2** contains 412 CRC 16S amplicon samples processed by Parallel-Meta Suite.

**Real Dateset 3** contains 139 CRC WGS Metagenome samples processed by MetaPhlAn4.

**Real Dateset 4** contains 104 T2D WGS Metagenome samples processed by MetaPhlAn4.

## Contact

All problems please contact Meta-Spec development team: **Xiaoquan Su**  Email: [suxq@qdu.edu.cn](mailto:suxq@qdu.edu.cn)
